lib/winapi/filemanip.c

```diff
@@ -241,6 +241,138 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
     }
 }
 
+BOOL CopyFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName, BOOL bFailIfExists)
+{
+    NTSTATUS status;
+    HANDLE sourceHandle;
+    HANDLE targetHandle = INVALID_HANDLE_VALUE;
+    ANSI_STRING sourcePath;
+    ANSI_STRING targetPath;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_NETWORK_OPEN_INFORMATION networkOpenInformation;
+    LPVOID readBuffer = NULL;
+    SIZE_T readBufferRegionSize = 0x1000;
+    DWORD bytesRead;
+
+    RtlInitAnsiString(&sourcePath, lpExistingFileName);
+    InitializeObjectAttributes(&objectAttributes, &sourcePath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(
+            &sourceHandle,
+            FILE_GENERIC_READ,
+            &objectAttributes,
+            &ioStatusBlock,
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+        sourceHandle,
+        &ioStatusBlock,
+        &networkOpenInformation,
+        sizeof(networkOpenInformation),
+        FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
```

Comment:
Sorry, I should've been more clear about what I meant. I meant to suggest you use `CloseHandle` instead of `NtClose` where we're interested in the error value, since `CloseHandle` already converts the `NTSTATUS` and calls `SetLastError`, but where we don't care about that (or want to preserve the previous error), a call to `NtClose` is better.

Reply:
Gotcha. 
I'd be a little worried that mixing the two may look like a bug/mistake if the reader does not know the nuanced difference between the behavior around setting `LastError` (like me, with very rusty memory about win32). I'd personally be inclined to use the more explicit error handling across the board (i.e., stick with `NtClose`) despite the slight duplication just for the sake of consistency, but defer to you.

Reply:
(I'll wait to make sure we're aligned before changing one way or the other and will wait to squash changes again until it's done so comments don't get orphaned while still under review)

Reply:
@thrimbor - just wanted to make sure you saw my comment above. Happy to go in whatever direction you prefer!

Reply:
Sounds good, I think it may be too nuanced to split usage and worry someone might "fix" it in the future, potentially introducing changes to the `LastError` behavior.

Reply:
Changed.